consensus/istanbul/core/prepare.go

```diff
@@ -19,49 +19,124 @@ package core
+		}
+
+		if signer != message.Address {
+			return errInvalidPreparedCertificateMsgSignature
+		}
+
+		// Check for duplicate messages
+		if seen[signer] {
+			return errInvalidPreparedCertificateDuplicate
+		}
+		seen[signer] = true
+
+		// Check that the message is a PREPARE or COMMIT message
+		if message.Code != istanbul.MsgPrepare && message.Code != istanbul.MsgCommit {
+			return errInvalidPreparedCertificateMsgCode
+		}
+
+		var subject *istanbul.Subject
+		if err := message.Decode(&subject); err != nil {
+			logger.Error("Failed to decode message in PREPARED certificate", "err", err)
+			return err
+		}
+
+		// Verify message for the proper sequence.
+		if subject.View.Sequence.Cmp(c.currentView().Sequence) != 0 {
+			return errInvalidPreparedCertificateMsgView
+		}
+
+		// Verify message for the proper proposal.
+		if subject.Digest != preparedCertificate.Proposal.Hash() {
+			return errInvalidPreparedCertificateDigestMismatch
+		}
+
+		// If COMMIT message, verify valid committed seal.
+		if message.Code == istanbul.MsgCommit {
+			_, src := c.valSet.GetByAddress(signer)
+			err := c.verifyCommittedSeal(subject.Digest, message.CommittedSeal, src)
+			if err != nil {
+				logger.Error("Commit seal did not contain signature from message signer.", "err", err)
+				return err
+			}
+		}
+	}
+
+	c.current.SetPreparedCertificate(preparedCertificate)
+	return nil
+}
+
+func (c *core) handlePrepare(msg *istanbul.Message) error {
+	logger := c.logger.New("state", c.state, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "handlePrepare", "tag", "handleMsg")
 	// Decode PREPARE message
 	var prepare *istanbul.Subject
 	err := msg.Decode(&prepare)
 	if err != nil {
 		return errFailedDecodePrepare
 	}
 
-	if err := c.checkMessage(msgPrepare, prepare.View); err != nil {
+	if err := c.checkMessage(istanbul.MsgPrepare, prepare.View); err != nil {
 		return err
 	}
 
-	// If it is locked, it can only process on the locked block.
-	// Passing verifyPrepare and checkMessage implies it is processing on the locked block since it was verified in the Preprepared state.
-	if err := c.verifyPrepare(prepare, src); err != nil {
+	if err := c.verifyPrepare(prepare); err != nil {
 		return err
 	}
 
-	c.acceptPrepare(msg, src)
+	c.acceptPrepare(msg)
+	logger.Trace("Accepted prepare", "Number of prepares or commits", c.current.GetPrepareOrCommitSize())
 
-	// Change to Prepared state if we've received enough PREPARE messages or it is locked
-	// and we are in earlier state before Prepared state.
-	if ((c.current.IsHashLocked() && prepare.Digest == c.current.GetLockedHash()) || c.current.GetPrepareOrCommitSize() >= c.valSet.MinQuorumSize()) &&
-		c.state.Cmp(StatePrepared) < 0 {
-		c.current.LockHash()
+	// Change to Prepared state if we've received enough PREPARE messages and we are in earlier state
+	// before Prepared state.
+	if (c.current.GetPrepareOrCommitSize() >= c.valSet.MinQuorumSize()) && c.state.Cmp(StatePrepared) < 0 {
```

Comment:
Consider calling `c.current.GetPrepareOrCommitSize()` and `c.valSet.MinQuorumSize()` only once